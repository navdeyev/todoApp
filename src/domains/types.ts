import {RouterState} from 'react-router-redux';

import {ITheme} from 'components/Styled/themes';
import {MediaType} from 'components/Styled/utils';
import {LoadingStates} from 'domains/loadingStates';
import {ITodo} from 'domains/todos/todosTypes';
import {IWindowDimensions} from 'domains/window/windowTypes';

export interface ITodosState {
    loadingState: LoadingStates,
    todos: ITodo[],
    selectedTodoId: string
}

export interface IWindowState {
    windowDimensions: IWindowDimensions,
    media: MediaType
}

export interface IThemeState {
    theme: ITheme
}

export interface IAppState {
    routing: RouterState,
    todosState: ITodosState,
    windowState: IWindowState,
    themeState: IThemeState
}
