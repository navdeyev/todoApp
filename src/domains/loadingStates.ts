export enum LoadingStates {
    NOT_STARTED = 'NOT_STARTED',
    LOADING = 'LOADING',
    COMPLETE = 'COMPLETE',
}