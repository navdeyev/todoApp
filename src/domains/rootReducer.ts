import {routerReducer} from 'react-router-redux';
import {combineReducers} from 'redux';

import {IAppState} from './types';

import todosReducer from 'domains/todos/todosReducer';
import windowReducer from 'domains/window/windowReducer';

const rootReducer = combineReducers<IAppState>({
    routing: routerReducer,
    todosState: todosReducer,
    windowState: windowReducer
});

export default rootReducer;