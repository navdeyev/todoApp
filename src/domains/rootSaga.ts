import {all, fork} from 'redux-saga/effects';

import todosSaga from './todos/todosSaga';

import {IServiceMap} from 'service/services';

function* rootSaga(serviceMap: IServiceMap) {
    yield all([
        fork(todosSaga, serviceMap)
    ]);
}

export default rootSaga;