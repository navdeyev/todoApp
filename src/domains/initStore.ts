import {applyMiddleware, createStore, Store} from 'redux';
import {composeWithDevTools} from 'redux-devtools-extension'
import thunk from 'redux-thunk';

import rootReducer from './rootReducer';
import {IAppState} from './types';

const initStore = (): Store<IAppState> => {
    const middlewares = [
        thunk
    ];

    const enhancer = process.env.NODE_ENV === 'development'
        ? composeWithDevTools(applyMiddleware(...middlewares))
        : applyMiddleware(...middlewares);

    return createStore<IAppState>(rootReducer, enhancer);
};

export default initStore;