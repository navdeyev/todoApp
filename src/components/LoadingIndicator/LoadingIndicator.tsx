import * as React from 'react';
import styled from 'styled-components';

import {LoadingStates} from 'domains/loadingStates';

import logo from './logo.svg';

export interface IProps {
    loadingState: LoadingStates,
    children: React.ReactChildren | React.ReactChild
}

const Outer = styled.div`
    position: relative;
    min-height: 100px;
`;
Outer.displayName = 'Outer';

const Inner = styled.div``;
Inner.displayName = 'Inner';

interface ICoverProps {
    visible: boolean;
}

export const Cover = styled.div`
    width: 100%;
    height: 100%;
    opacity: 0.5;
    background: #fff;
    position: absolute;
    align-items: center;
    justify-content: center;
    display: ${ (props: ICoverProps) => props.visible ? 'flex' : 'none' }
`;
Cover.displayName = 'Cover';

const Logo = styled.img`
  animation: App-logo-spin infinite 20s linear;
  height: 80px;
  
  @keyframes App-logo-spin {
    from { transform: rotate(0deg); }
    to { transform: rotate(360deg); }
  }
`;
Logo.displayName = 'Logo';

const LoadingIndicator: React.SFC<IProps> = (props) => {
    return (
        <Outer>
            <Cover visible={LoadingStates.LOADING === props.loadingState}>
                <Logo src={logo}/>
            </Cover>
            <Inner>
                {props.children}
            </Inner>
        </Outer>
    );
};

export default LoadingIndicator;