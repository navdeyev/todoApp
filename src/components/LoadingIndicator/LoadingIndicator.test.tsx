import {shallow} from 'enzyme';
import 'jest-styled-components';
import * as React from 'react';

import {LoadingStates} from 'domains/loadingStates';

import LoadingIndicator, {Cover, IProps} from './LoadingIndicator';

describe('Loading indicator', () => {
    let defaultProps: IProps;
    beforeEach(() => {
        defaultProps = {
            children: <div>Some content here</div>,
            loadingState: LoadingStates.COMPLETE
        };
    });

    it('hides Cover if loading state is LoadingStates.COMPLETE', () => {
        const render = shallow(<LoadingIndicator {...defaultProps} />);
        expect(render.find(Cover).props().visible).toBe(false);
        expect(render.getElement()).toMatchSnapshot();
    });

    it('shows Cover if loading state is LoadingStates.LOADING', () => {
        defaultProps.loadingState = LoadingStates.LOADING
        const render = shallow(<LoadingIndicator {...defaultProps} />);
        expect(render.find(Cover).props().visible).toBe(true);
        expect(render.getElement()).toMatchSnapshot();
    });

});

describe('Cover', () => {
    it('applies display: block style for visible state', () => {
        const render = shallow(<Cover visible={true}/>);
        expect(render).toHaveStyleRule('display', 'flex');
    });

    it('applies display: none style for hidden state', () => {
        const render = shallow(<Cover visible={false}/>);
        expect(render).toHaveStyleRule('display', 'none');
    });
});