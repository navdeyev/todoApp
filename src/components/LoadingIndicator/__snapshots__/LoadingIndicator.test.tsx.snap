// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading indicator hides Cover if loading state is LoadingStates.COMPLETE 1`] = `
<Outer>
  <Cover
    visible={false}
  >
    <Logo
      src={undefined}
    />
  </Cover>
  <Inner>
    <div>
      Some content here
    </div>
  </Inner>
</Outer>
`;

exports[`Loading indicator shows Cover if loading state is LoadingStates.LOADING 1`] = `
<Outer>
  <Cover
    visible={true}
  >
    <Logo
      src={undefined}
    />
  </Cover>
  <Inner>
    <div>
      Some content here
    </div>
  </Inner>
</Outer>
`;
